### PR TITLE
Switches!

### PR DESCRIPTION
**Pull request by mikeday**
(documentation to go here)

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -237,315 +241,277 @@ impl Format {
...
         }
     }
-    /*
-        pub fn disjoint(a: &Self, b: &Self) -> bool {
-            for i in 0..std::cmp::min(a.pattern.len(), b.pattern.len()) {
-                if ByteSet::disjoint(&a.pattern[i], &b.pattern[i]) {
-                    return true;
-                }
+
+    fn add_next(&mut self, index: usize, depth: usize, next: &Next) -> Result<(), ()> {
```
Comment (brendanzab) [inline src/main.rs:274]: Is the `depth` the number of bytes of lookahead?

File: src/main.rs [hunk 2]
```diff
@@ -94,14 +96,15 @@ enum Format {
...
 }
 
-#[derive(Debug)]
-struct Lookahead {
-    pattern: Vec<ByteSet>,
+#[derive(Clone, Debug)]
+struct Switch {
+    accept: Option<usize>,
+    branches: Vec<(ByteSet, Switch)>,
 }
```
Comment (brendanzab) [inline src/main.rs:103]: Is this sort of like a tree of lookahead? Keeping track of the index of variant matched in the `accept` field?

File: src/main.rs [hunk 3]
```diff
@@ -237,315 +241,277 @@ impl Format {
...
+
+    fn build(branches: &[&Format], next: &Next) -> Option<Switch> {
+        const MAX_DEPTH: usize = 16;
+        for depth in 0..MAX_DEPTH {
+            if let Some(switch) = Switch::build0(depth, branches, next) {
+                return Some(switch);
+            }
+        }
+        None
+    }
```
Comment (brendanzab) [inline src/main.rs:402]: I imagine this is trying `0..MAX_DEPTH` bytes of lookahead, trying a longer lookahead on failure?

File: src/main.rs (hunk 1 above)
Comment (mikeday) [reply src/main.rs:274]: Yes, each level of the tree switches on the next byte.

File: src/main.rs (hunk 2 above)
Comment (mikeday) [reply src/main.rs:103]: Yep! If none of the (disjoint) bytesets match or if there is no input remaining then the accept field indicates which if any of the switch cases succeeds.

File: src/main.rs (hunk 3 above)
Comment (mikeday) [reply src/main.rs:402]: Yes, we want to construct the shallowest possible tree as deeper trees can take a very long time to construct on some pathological input. There may still be inputs that exhibit awful time complexity, unfortunately.

**Pull request merged**